BENCH-1228 Add updated data collections notebook

emmarogge opened the pull request:
Refactor data collections notebook to guide users through creation of a new workspace, conversion to data collection, and publication of versions.

emmarogge committed:
BENCH-1228 Add updated data collections notebook

Refactor data collections notebook to guide users through creation of a new workspace, conversion to data collection, and publication of versions.

wnojopra reviewed (changes requested):
The working_with_data_collections.ipynb notebook isn't loading in both github and on jupyter for me. Please fix
```
File Load Error for working_with_data_collections.ipynb
Unreadable Notebook: /home/jupyter/repos/terra-axon-examples/first_hour_on_vwb/working_with_data_collections.ipynb NotJSONError('Notebook does not appear to be JSON: \'{\\n "cells": [\\n {\\n "cell_type": "m...')
```

emmarogge committed:
Remove metadata saved from ipywidgets

emmarogge commented:
> The working_with_data_collections.ipynb notebook isn't loading in both github and on jupyter for me. Please fix
> 
> ```
> File Load Error for working_with_data_collections.ipynb
> Unreadable Notebook: /home/jupyter/repos/terra-axon-examples/first_hour_on_vwb/working_with_data_collections.ipynb NotJSONError('Notebook does not appear to be JSON: \'{\\n "cells": [\\n {\\n "cell_type": "m...')
> ```

I have now fixed this.

wnojopra commented on first_hour_on_vwb/working_with_data_collections.ipynb:
The actual code in the notebook is focused on creation, with brief instruction at the end on how others can add it. So I think we should rename this to "Creating a data collection".

emmarogge replied:
done!

wnojopra commented on first_hour_on_vwb/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,603 @@
+   "source": [
+    "## Overview\n",
+    "\n",
+    "This notebook provides a walkthrough of creating and using a [data collection](https://terra-docs.api.verily.com/docs/reference/glossary/#data-collection) in Verily Workbench. A data collection is a grouping of related cloud-based resources related to a specific project, study or purpose. To interact with the data in a data collection, you must have access at the policy level, via group memb ...
```
Assuming we reframe this notebook as a "Creating a Data Collection", we should adjust this section appropriately.

emmarogge replied:
done!

wnojopra commented on first_hour_on_vwb/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,603 @@
+    "tags": []
+   },
+   "source": [
+    "### Create a workspace\n",
```
```suggestion
    "### Create a new workspace for the data collection\n",
```
Suggest making it obvious that its a new workspace users need to create, to help with "I thought already created a workspace" confusion

emmarogge replied:
done!

wnojopra commented on first_hour_on_vwb/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,603 @@
+    "<a id=\"create-new-dc\"></a>\n",
+    "\n",
+    "Before creating your data collection, consider the following:\n",
+    "1. What data do you want to share? What type of resources--Cloud Storage buckets or objects, BigQuery tables, GitHub repositories--will be made available via this data collection?\n",
```
```suggestion
    "1. What data do you want to share? What type of resources--Cloud Storage buckets or objects, BigQuery tables--will be made available via this data collection?\n",
```
I think based on some older implementation, github repos were able to be a part of data collections. I don't _think_ thats the case anymore, and I don't think we intend to have repos be part of a data collection.

amygdala replied:
Harris confirmed that this is not the case any more.

wnojopra replied:
Sorry @amygdala can you please clarify what is not the case? Is the statement "Github repos are not a part of data collections" true?

amygdala replied:
sorry!  your statement is true: github repos can no longer be part of data collections.
(this changed with the version support)

emmarogge replied:
fixed!

wnojopra commented on first_hour_on_vwb/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,603 @@
+    "\n",
+    "Before creating your data collection, consider the following:\n",
+    "1. What data do you want to share? What type of resources--Cloud Storage buckets or objects, BigQuery tables, GitHub repositories--will be made available via this data collection?\n",
+    "1. With whom do you wish to share this data? Will you be sharing the data collection with all members of an existing Verily Workbench group (e.g. for your organization or team), or will you need to create a new Verily Workbench group in order to restrict access to the data collection?\n",
```
```suggestion
    "1. With whom do you wish to share this data? Will you be sharing the data collection with all members of an existing Workbench group (e.g. for your organization or team), or will you need to create a new Workbench group in order to restrict access to the data collection?\n",
```
I think Sean mentioned we should favor `Workbench` over `Verily Workbench` as branding guidance. Please do a search and replace in this file.

emmarogge replied:
done!

wnojopra commented on first_hour_on_vwb/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,603 @@
+    "\n",
+    "<div class=\"alert alert-block alert-success\">\n",
+    "<b>Note:</b> \n",
+    "    If you'd like to restrict access to your data collection to members of a specific group, contact <a href=\"mailto:workbench-support@verily.com\">Verily Workbench Support</a> <b>before</b> creating your workspace so that a <a href=\"https://terra-docs.api.verily.com/docs/https://terra-docs.api.verily.com/docs/reference/glossary/#policy\">group policy constraint</a> can be applied at the t ...
```
I don't believe we need to guide users to contact support anymore for this. You can now set the policies for the data collection’s backing workspace at creation time through the UI. Can we also set workspace policies through the UI? If yes, we should update the notebook to do so. If no, we should guide users to create the workspace with policy through the UI. And then file a ticket to add CLI support for workspace policy.

wnojopra replied:
I just realized that what I said above applies only for **group** policies. For **regional** policies, we'd still need to be contacted.

emmarogge replied:
That is correct. I think, therefore, we should still include the link to support here in addition to instructions for adding the group membership constraint in the UI.

amygdala commented on first_hour_on_vwb/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,603 @@
+   "metadata": {
+    "tags": []
+   },
+   "source": [
```
I think that ideally we want some kind of check that the workspace has finished creating, before they move on to the next section. (Else I assume things will not go well).  This could easily happen if someone just mindlessly proceeds from that cell to the next.

The check could be another terra-cli command, or potentially they could visit their list of workspaces in the UI.

amygdala replied:
Thinking about this further, a few comments:
- this is not quite as error-prone as I initially thought: even though the widget cell looks like it's done running ( it has no `[*]` indication), apparently the next cell won't run until it's finished.
- We do still need to worry about someone mindlessly running one cell after the other, since the dropdowns aren't set to the DC workspace
- So I think we could address this as mentioned in a subsequent comment: **let's set the focus in the subsequent dropdowns to the data collection workspace they just created**.  The notebook has that info, so hopefully it would be relatively straightforward to do so.  
That way, mindlessly clicking 'next' will do basically the right thing.

emmarogge replied:
In the initial version of this notebook, I did have each widget that follows the workspace creation one auto-populate with the newly created workspace's ID. However, it occurred to me that users might want to use the widgets out of order, in which case auto-population might be annoying/might prevent them from realizing they can use the conversion and version-publishing widgets for existing workspaces without creating a new WS. (You are usually the person who reminds me that users may not proceed in a linear fashion!)

One specific use case where this notebook instructs the user to run the widgets out of order is for those who want to add a discovery constraint. Those users skip the "Create a workspace" widget entirely and need to select a workspace from the dropdown.

It seems like it might work for both use cases (those who mindlessly run every cell in linear order AND those who want to use some subset of the widgets out of order)  to create a section of the notebook for converting existing workspaces to data collections and another section for publishing new versions of existing data collections. This will make the notebook substantially longer and essentially be duplicating existing code, for the most part, increasing the maintenance burden. What do you think?

amygdala replied:
I think that's over-complicating :)
I'd say: if the new DC var is set, set the focus of the dropdown to that value; otherwise leave the focus unset (i.e., don't just set to the first in the list) so the user needs to make an explicit choice.

Or, if that seems too complex, just do the latter (require the user to make an explicit choice, don't pre-set to any workspace).

amygdala commented on first_hour_on_vwb/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,603 @@
+   },
+   "outputs": [],
+   "source": [
+    "class ConvertToDataCollectionWidget(object):\n",
```
The pulldown of workspace IDs is nice, but I think we ought to try to auto-set it to the just-created data collection workspace (we should be able to collect that ID from running the previous widget). Else, have the pulldown start out with no item selected so that they need to explicitly choose.

Otherwise, I'm worried that someone (mindlessly clicking) will set the data collection property on the random first workspace in the list.

emmarogge replied:
See my proposal in previous comment (:

amygdala replied:
I replied there.

emmarogge committed:
Update to address reviewer feedback.

amygdala commented on first_hour_on_vwb/working_with_data_collections.ipynb:
```diff
@@ -0,0 +1,603 @@
+    "### Convert new workspace to data collection\n",
+    "<a id=\"convert-to-dc\"></a>\n",
+    "\n",
+    "Now you'll convert your newly created workspace, to which you have added resources, into a data collection which can be shared with others and added to other workspaces. \n",
```
I think this section needs a bit on that 'converting to a data collection' doesn't yet mean that they will see their new DC in the data catalog yet— it won't show up there until they publish a version, which they'll do next.

emmarogge replied:
fixed!

amygdala commented on first_hour_on_vwb/creating_a_data_collection.ipynb:
```diff
@@ -0,0 +1,630 @@
+   },
+   "outputs": [],
+   "source": [
+    "class PublishVersionWidget(object):\n",
```
I think this widget (**and maybe others**) need to quote the description string, I'm getting this error:
 
<img width="760" alt="CleanShot 2023-09-22 at 08 52 26@2x" src="https://github.com/DataBiosphere/terra-axon-examples/assets/115093/1f885781-bd08-481c-b647-f40ef83b8825">

emmarogge replied:
Fixed!

amygdala commented on first_hour_on_vwb/creating_a_data_collection.ipynb:
```diff
@@ -465,9 +492,9 @@
     "<a id=\"add-dc-to-ws\"></a>\n",
     "\n",
-    "Follow the steps below to add your new data collection to an Verily Workbench workspace.<br>The video below provides a visual walkthrough of these steps.\n",
+    "Follow the steps below to add your new data collection to an Workbench workspace.<br>The video below provides a visual walkthrough of these steps.\n",
```
nit: "an Workbench" --> "a Workbench"

emmarogge replied:
fixed!

amygdala commented on first_hour_on_vwb/creating_a_data_collection.ipynb:
```diff
@@ -0,0 +1,630 @@
+    "### Add data collection to workspace\n",
+    "<a id=\"add-dc-to-ws\"></a>\n",
+    "\n",
+    "Follow the steps below to add your new data collection to an Workbench workspace.<br>The video below provides a visual walkthrough of these steps.\n",
```
It feels like at the end of this section, we should direct them to the UI where they can see that if they click on 'data catalog' they can see what their new DC & version will look like to users.

emmarogge replied:
I believe the embedded screencast does just that.

amygdala replied:
IMO it would be helpful to add another line (that links to the UI) for those who prefer text to video-watching-- but your call.

amygdala commented on first_hour_on_vwb/creating_a_data_collection.ipynb:
```diff
@@ -0,0 +1,630 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Add resources to a data collection version\n",
```
IMO it would be helpful here to have them create a resource in _this_ notebook (e.g. just a controlled bucket, or maybe a referenced resource pointing to something public), and not make them go elsewhere to do this step.  Then add a command to move _that_ resource to the version folder.  

This would have 2 benefits:
- don't make them break the flow by going elsewhere
- the important step to get the resource into the right folder can be runnable, not just markdown that their eyes might glaze over

emmarogge replied:
I am planning to do this in a subsequent PR -- I have built the tooling to actually move ALL the contents of a given folder to the version folder, but that section would benefit from its own review. Resolving accordingly.

amygdala commented:
A few high-level comments:
- it now makes sense to link somewhere in this nb to the doc on this now it's out, e.g. https://support.workbench.verily.com/docs/how_to_guides/creating_data_collections/#how-to-create-a-data-collection-and-manage-its-versions 
- Somewhat related, I think the discussion of policies could use a bit of rework.  You might want to link here: https://support.workbench.verily.com/docs/how_to_guides/creating_data_collections/#add-a-data-collection-policy
Basically, they can set group policies themselves via the web UI, ideally **before** they share the workspace, but need to contact us to set region policies. 
- You should also clarify (or point to docs) regarding the difference between setting a group policy, and sharing the workspace with users or groups, as they're 2 diff things.
Regardless of whether there's a group policy, you probably want to spell out that once they've created a dc, they need to share the underlying workspace with the right target audience so that they can see the dc. (sorry if I missed that).

emmarogge committed:
Addressed reviewer feedback

emmarogge committed:
Fix typo

emmarogge committed:
Remove unused lines of code from utility file

emmarogge commented:
> A few high-level comments:
> 
> * it now makes sense to link somewhere in this nb to the doc on this now it's out, e.g. https://support.workbench.verily.com/docs/how_to_guides/creating_data_collections/#how-to-create-a-data-collection-and-manage-its-versions
I have linked to this doc in the notebook section "Restricting discovery access to a data collection"
> * Somewhat related, I think the discussion of policies could use a bit of rework.  You might want to link here: https://support.workbench.verily.com/docs/how_to_guides/creating_data_collections/#add-a-data-collection-policy
I refactored the content about policies and linked to the indicated guide
>   Basically, they can set group policies themselves via the web UI, ideally **before** they share the workspace, but need to contact us to set region policies.
done
> * You should also clarify (or point to docs) regarding the difference between setting a group policy, and sharing the workspace with users or groups, as they're 2 diff things.
done
>   Regardless of whether there's a group policy, you probably want to spell out that once they've created a dc, they need to share the underlying workspace with the right target audience so that they can see the dc. (sorry if I missed that).
done

amygdala commented on first_hour_on_vwb/creating_a_data_collection.ipynb:
```diff
@@ -0,0 +1,735 @@
+   },
+   "outputs": [],
+   "source": [
+    "class CreateWorkspaceWidget(object):\n",
```
I noticed that with the 'create workspace' widget, the description field content, e.g. as shown here:
<img width="791" alt="CleanShot 2023-09-27 at 11 04 07@2x" src="https://github.com/DataBiosphere/terra-axon-examples/assets/115093/e8a77ed1-337f-4994-a97b-46da9867aa7b">
ends up with quotes around it in the created workspace description:
<img width="551" alt="CleanShot 2023-09-27 at 11 04 21@2x" src="https://github.com/DataBiosphere/terra-axon-examples/assets/115093/09394f3f-f2e4-4ed3-bbf0-73fb9a3d3a9c">


The user can go in to the UI and edit that, but it would be nice for them to not have to do that.
However, this might not be fixable... 

Update: when I tested from the command line using _single_ quotes around the description, the result was not quoted.

emmarogge replied:
I agree that having to manually remove delimiters via the UI  is annoying; however, when I tested with single quotes, the single quotes still appeared (with CLI v0.411.0). I have therefore changed the widget such that all leading and trailing whitespace is stripped but the string is undelimited, which does not add delimiters to the description field.

amygdala commented on first_hour_on_vwb/creating_a_data_collection.ipynb:
```diff
@@ -0,0 +1,735 @@
+   },
+   "outputs": [],
+   "source": [
+    "class ConvertToDataCollectionWidget(object):\n",
```
When I run this widget, it now throws an error for me: https://screenshot.googleplex.com/BHE6VgKR2ME47se

emmarogge replied:
Unfortunately, even when I zoom in, the error message in that screenshot is illegible as it is so small. Could you please link to a Gpaste of the specific error? I have not been able to replicate any error in running that widget.

emmarogge committed:
Pushing changes to test with new CLI command outside of cloud env

emmarogge committed:
Address reviewer feedback

Remove characters delimiting data collection description so that they do not appear in UI. Refactor utilities file due to [implementation of JSON-formatted output of workspace folder tree via Workbench CLI](https://github.com/verily-src/terra-tool-cli/pull/319). Add pytest unit tests for utility methods.

emmarogge committed:
Populate second widget's dropdown automatically

In the 'convert to data collection' widget, automatically populate the workspace ID of the new workspace created in the previous 'create workspace' widget/step.

emmarogge committed:
Merge remote-tracking branch 'origin/main' into emmarogge/BENCH-1228

emmarogge committed:
Remove content about switching workspaces

It is not necessary to switch away from workspace created and converted to a data collection to perform the subsequent operations.

wnojopra commented on first_hour_on_vwb/creating_a_data_collection.ipynb:
```diff
@@ -0,0 +1,709 @@
+    "        self.label = widgets.Label(value = 'Please provide appropriate values in the input boxes.')\n",
+    "        self.data_collection_ids = self.get_data_collections()\n",
+    "        if prev_widget and prev_widget.get_workspace_id() != '':\n",
+    "            self.input_workspace_id = wu.DropdownInputWidget(self.data_collection_ids,prev_widget.get_workspace_id(),\"Workspace ID:\").get()\n",
```
I got an error at this point saying `ValueError: 'willyn-workspace-dc-test' not in array`.

It turns out this wasn't your fault, but `terra workspace list` has a default limit of 30. `willyn-workspace-dc-test` happened to be my 31st workspace. Might be worth adding `--limit=100` or something to the `terra workspace list` call.

emmarogge replied:
Now that I've refactored this (per our Slack convo yesterday), this comment is obsolete.

wnojopra commented on first_hour_on_vwb/creating_a_data_collection.ipynb:
```diff
@@ -0,0 +1,709 @@
+    "            print(setWorkspaceResult.stderr) if not setWorkspaceResult.stdout else print(setWorkspaceResult.stdout)\n",
+    "\n",
+    "            # 2) Get file tree from Terra CLI.\n",
+    "            tree_dictionary = vfu.parse_tree()\n",
```
At this point I got an error `AttributeError: module 'vwb_folder_utils' has no attribute 'parse_tree'`

emmarogge replied:
Fixed -- this was an oversight on my part.

emmarogge committed:
Fix bug in last two widgets

Refactor last two widgets to utilize correct helper method.

emmarogge committed:
Automatically use previous widget's workspace ID

Use the workspace ID from the previous widget.

emmarogge committed:
Fix sections' parameter descriptions so they are parallel

emmarogge committed:
Merge branch 'main' into emmarogge/BENCH-1228

wnojopra commented on first_hour_on_vwb/test_vwb_folder_utils.py:
This file isn't meant to be run or imported by the user, right? Lets move this to a new `test` folder under `first_hour_on_vwb`

emmarogge replied:
Roger!

wnojopra reviewed (approved):
Nice!! Everything worked out well for me. Just one nit on the test and I think this is good to merge.

emmarogge committed:
Add data collections notebook to README

Add creating_a_data_collection.ipynb description and hyperlink to README. Move test file to a new test directory under first_hour_on_vwb/ directory.
